posts/inside-rust/2023-11-10-parallel-rustc.md

```diff
@@ -0,0 +1,239 @@
+  its internal code representation (MIR) to LLVM's code representation (LLVM
+  IR) in serial. This takes a brief period for each codegen unit, and explains
+  the staircase shape on the left-hand side of the code generation threads.
+  There is some room for improvement here.
+- The front-end is entirely serial. There is a lot of room for improvement
+  here.
+
+### New intraprocess parallelism: the front-end
+
+The front-end is now capable of parallel execution. It uses
+[Rayon](https://crates.io/crates/rayon) to perform compilation tasks using
+fine-grained parallelism. Many data structures are synchronized by mutexes and
+read-write locks, atomic types are used where appropriate, and many front-end
+operations are made parallel. The addition of parallelism was done by modifying
+a relatively small number of key points in the code. The vast majority of the
+front-end code did not need to be changed.
+
+When the parallel front-end is enabled and configured to use eight threads, we
+get the following Samply profile when compiling the same example as before.
+
+![Samply output when compiling Cargo, parallel](../../../../images/inside-rust/2023-11-10-parallel-rustc/samply-parallel.png)
+
+Again, there are several things worth nothing.
+- Front-end execution takes 5.9 seconds (down from 10.2 seconds).
+- Back-end execution takes 5.3 seconds (down from 6.2 seconds), and the LLVM
+  threads are running for 4.9 seconds of that (down from 5.9 seconds).
+- There are seven additional threads labelled "rustc" operating in the
+  front-end. The reduced front-end time shows they are reasonably effective,
+  but the thread utilization is patchy, with the eight threads all having
+  periods of inactivity. There is room for significant improvement here.
+- Eight of the LLVM threads start at the same time. This is because the eight
+  "rustc" threads create the LLVM IR for eight codegen units in parallel. (For
+  seven of those threads that is the only work they do in the back-end.) After
+  that, the staircase effect returns because only one "rustc" thread does LLVM
+  IR generation while seven or more other LLVM threads are active. If the
+  number of threads used by the front-end was changed to 16 the staircase shape
+  would disappear entirely, though in this case the final execution time would
+  barely change.
+
+### Putting it all together
+
+Rust compilation has long benefited from intraprocess parallelism, via Cargo,
+and from interprocess parallelism in the back-end. It can now also benefit from
+interprocess parallelism in the front-end.
+
+Attentive readers might be wondering how intraprocess parallelism and
+interprocess parallelism interact. If we have 20 parallel rustc invocations and
+each one can have up to 16 threads running, might we end up 100s of threads on
+a machine with only 10s of cores, resulting in inefficient execution as the OS
+tries its best to schedule them?
```

Comment:
I first read this as "one hundred seconds" thinking in CPU time, before realizing it meant "hundreds". This might be more clear:

```suggestion
Attentive readers might be wondering how intraprocess parallelism and
interprocess parallelism interact. If we have 20 parallel rustc invocations and
each one can have up to 16 threads running, might we end up with hundreds of threads on
a machine with only tens of cores, resulting in inefficient execution as the OS
tries its best to schedule them?
```